### DISCUSSION TITLE
Persisting Switch Profile target on profile export / import

### DISCUSSION THREAD

**Discussion by salimv** [elgatosf/elgato#27] [Help] [answered]
**Software**: 6.2.1 (18825)
**Hardware**: Streamdeck 62 (15 key)
**Issue**: Exporting profile with Switch Profile Navigation action to switch to another profile, does not seem to remember and reload the target profile upon import of the exported profile.

**Use Case**:
- We use custom Streamdeck plugin to activate certain actions of our web based app as hotkeys
- Our webapp has tabs, and each tab has a different set of actions
- The goal here is to show context relevant actions in the streamdeck based on the tab the user is on
- The approach to implement this is thru `switchToProfile` event. The web app would sent switchToProfile event to streamdeck thru the plugin to switch the profile to the desired profile.
- Each profile also has Switch Profile Navigation action buttons on it, to allow switching to other profiles directly from streamdeck

**Steps to reproduce:**
1) Create two Profiles - Profile A and Profile B
2) In Profile A, drag drop the Switch Profile Navigation. Set profile property of the action to Profile B. Repeat this in Profile B, pointing the switch profile target to Profile A
3) Test to make sure it works
4) Export Profile A and Profile B
5) Delete Profile A and Profile B from the App
6) Import Profile A and Profile B from the exported file
7) Check the profile property. I see it is always reset to 'Default' profile in both profiles. 

I suspect the app stores some internal ID representing the profile as button target and not the name. Hence, when the profiles are deleted, those ID become invalid. Reimporting creates new IDs, and since the old IDs no longer exist, the button profile target properties are set to Default.

**Comment by ChekTek** [comment, marked as answer]
Hi salimv!

One way you can work around this is by creating your own "Switch To Profile" action instead of using Stream Deck's built-in Switch Profile action (Note that these only work for profiles that are bundled within the plugin for now). Your plugin's "Switch To Profile" action would use the `keyUp` event to send a `switchToProfile` event to Stream Deck.

<https://docs.elgato.com/sdk/plugins/events-sent#switchtoprofile>

If you only want the action to be available in the profile and not visible in the action list, you can use the `VisibleInActionsList` flag in your manifest.json to hide it.

<https://docs.elgato.com/sdk/plugins/manifest#members>

I hope this helps. Let me know if you have any questions!

I will also discuss this with the team to see if there may be a better way to keep track of Switch Profile keys inside of profiles.

**Comment by salimv** [comment]
Thank you @ChekTek. This is helpful. We did take the same route to implement our own `Switch To Profile` action. Your note to use the `VisibleInActionsList` is a valuable pointer. 

I have a related followup question for helping with customer upgrades. I now have 5+ profiles to switch between. When the customer re-installs or upgrades the plugin, I have not been able to find a way to also replace the included profiles. The re-installation/upgrade process ends up adding copies of the profile, which means customer have to manually delete all profiles before the re-installation/upgrade. Is there a way to also replace the profiles instead of creating copies every time the plugin is re-installed or upgraded?

**Reply by ChekTek** [reply]
Unfortunately, we do not have a way to update installed profiles at this time, but I will talk to the team about this feature request for the potential future.

**Reply by MillerMark** [reply]
I have a similar usability challenge, where my plugin includes nine different profiles. There is no easy way to delete all nine of them when testing an install or when upgrading from an old version to a new version (I also had to implement my own "Switch to Profile" button).

**Reply by salimv** [reply]
Hello @ChekTek @elgatosystems  - Do we have any further development on the issue that re-install or upgrade of plugin, that includes profiles, does not replace those profiles but creates new sets of copies for each profile ? This is a pain during development/testing as well as when customers upgrade the plugin.

**Reply by ChekTek** [reply]
Hi @salimv 

There aren't plans to change the profile system at this time, however, we are working on something we are calling "Full Canvas Control".  Full Canvas Control, will allow a single action take over the entire Stream Deck canvas, similar to a plugin with a built-in profile. This should be easier for developers to implement and also provide a better user experience.
